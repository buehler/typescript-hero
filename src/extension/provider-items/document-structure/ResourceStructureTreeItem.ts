import { ExtensionContext, TreeItemCollapsibleState } from 'vscode';

import { Resource } from '../../../common/ts-parsing/resources';
import { BaseStructureTreeItem } from './BaseStructureTreeItem';
import { DeclarationStructureTreeItem } from './DeclarationStructureTreeItem';
import { ImportsStructureTreeItem } from './ImportsStructureTreeItem';

/**
 * Structure item that represents an additional resource in a file (namespace etc).
 * 
 * @export
 * @class ResourceStructureTreeItem
 * @extends {BaseStructureTreeItem}
 */
export class ResourceStructureTreeItem extends BaseStructureTreeItem {
    constructor(public resource: Resource, private context: ExtensionContext) {
        super(resource.identifier);
        this.collapsibleState = TreeItemCollapsibleState.Collapsed;
        this.iconPath = this.context.asAbsolutePath('./src/extension/assets/icons/declarations/module.svg');
    }

    public getChildren(): BaseStructureTreeItem[] {
        const items: BaseStructureTreeItem[] = [];
        if (this.resource.imports && this.resource.imports.length) {
            items.push(new ImportsStructureTreeItem(this.resource, this.context));
        }
        items.push(...this.resource.resources.map(r => new ResourceStructureTreeItem(r, this.context)));
        items.push(...this.resource.declarations.map(d => new DeclarationStructureTreeItem(d, this.context)));
        return items;
    }
}
