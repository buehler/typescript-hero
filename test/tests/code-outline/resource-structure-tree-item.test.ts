import { expect } from 'chai';
import * as snapshot from 'snap-shot-it';
import { File, Namespace, StringImport, VariableDeclaration } from 'typescript-parser';
import { ExtensionContext } from 'vscode';

import ResourceStructureTreeItem from '../../../src/code-outline/resource-structure-tree-item';
import ioc from '../../../src/ioc';
import iocSymbols from '../../../src/ioc-symbols';

describe('ResourceStructureTreeItem', () => {

  let context: ExtensionContext;

  before(() => {
    context = ioc.get<ExtensionContext>(iocSymbols.extensionContext);
  });

  it('should create a tree item', () => {
    const resource = new File('./path', '/root', 0, 100);
    const item = new ResourceStructureTreeItem(resource, context);

    expect(item).to.exist;
  });

  it.skip('should return correct children', () => {
    const resource = new File('./path', '/root', 0, 100);
    resource.imports.push(new StringImport('str-imp'));
    resource.declarations.push(new VariableDeclaration('var', false, false, undefined));
    resource.resources.push(new Namespace('namespace', 10, 20));

    const item = new ResourceStructureTreeItem(resource, context);

    snapshot(item.getChildren());
  });

});
