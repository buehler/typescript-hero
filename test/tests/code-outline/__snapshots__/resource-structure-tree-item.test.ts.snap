// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceStructureTreeItem should return correct children 1`] = `
Array [
  Object {
    "ctor": "ImportsStructureTreeItem",
    "label": "Imports",
  },
  Object {
    "ctor": "ResourceStructureTreeItem",
    "label": "namespace",
  },
  Object {
    "ctor": "DeclarationStructureTreeItem",
    "label": "var",
  },
]
`;
